// @inworld/nodejs-sdk/build/src/services/gprc/state_serialization_client_grpc.service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateSerializationClientGrpcService = void 0;
const grpc_js_1 = require("@grpc/grpc-js");
const state_serialization_grpc_pb_1 = require("../../../../proto/ai/inworld/engine/v1/state_serialization_grpc_pb");
const state_serialization_pb_1 = require("../../../../proto/ai/inworld/engine/v1/state_serialization_pb");
const util_1 = require("util");
const config_1 = require("../../common/config");
const constants_1 = require("../../common/constants");
const helpers_1 = require("../../common/helpers");
const logger_1 = require("../../common/logger");
class StateSerializationClientGrpcService {
    constructor() {
        this.config = config_1.Config.getInstance();
        this.address = this.config.getHost();
        this.ssl = this.config.getSSL();
        this.client = new state_serialization_grpc_pb_1.StateSerializationClient(this.config.getHost(), this.ssl ? grpc_js_1.credentials.createSsl() : grpc_js_1.credentials.createInsecure(), Object.assign({}, helpers_1.grpcOptions));
        this.logger = logger_1.Logger.getInstance();
    }
    getSessionState(props) {
        return __awaiter(this, void 0, void 0, function* () {
            const { scene, sessionToken } = props;
            const workspace = constants_1.SCENE_PATTERN.exec(scene)[1];
            const name = `workspaces/${workspace}/sessions/${sessionToken.sessionId}`;
            const request = new state_serialization_pb_1.GetSessionStateRequest().setName(name);
            const metadata = this.getMetadata(sessionToken);
            const response = yield (0, util_1.promisify)(this.client.getSessionState.bind(this.client))(request, metadata);
            this.logger.debug({
                action: 'Get session state',
                data: {
                    address: this.address,
                    ssl: this.ssl,
                    metadata: metadata.toJSON(),
                    request: request.toObject(),
                    response: response.toObject(),
                },
            });
            return {
                state: response.getState_asB64(),
                creationTime: response.getCreationTime().toDate().toISOString(),
            };
        });
    }
    getMetadata(sessionToken) {
        const metadata = new grpc_js_1.Metadata();
        metadata.add('authorization', `${sessionToken.type} ${sessionToken.token}`);
        return metadata;
    }
}
exports.StateSerializationClientGrpcService = StateSerializationClientGrpcService;
