// @inworld/nodejs-sdk/build/src/entities/emotion_strength.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmotionStrength = exports.EmotionStrengthCode = void 0;
const packets_pb_1 = require("../../../proto/ai/inworld/packets/packets_pb");
var EmotionStrengthCode;
(function (EmotionStrengthCode) {
    EmotionStrengthCode["UNSPECIFIED"] = "UNSPECIFIED";
    EmotionStrengthCode["WEAK"] = "WEAK";
    EmotionStrengthCode["STRONG"] = "STRONG";
    EmotionStrengthCode["NORMAL"] = "NORMAL";
})(EmotionStrengthCode = exports.EmotionStrengthCode || (exports.EmotionStrengthCode = {}));
class EmotionStrength {
    constructor(strength) {
        this.code = strength;
    }
    static fromProto(code) {
        switch (code) {
            case packets_pb_1.EmotionEvent.Strength.UNSPECIFIED:
                return EmotionStrengthCode.UNSPECIFIED;
            case packets_pb_1.EmotionEvent.Strength.WEAK:
                return EmotionStrengthCode.WEAK;
            case packets_pb_1.EmotionEvent.Strength.STRONG:
                return EmotionStrengthCode.STRONG;
            case packets_pb_1.EmotionEvent.Strength.NORMAL:
                return EmotionStrengthCode.NORMAL;
        }
    }
}
exports.EmotionStrength = EmotionStrength;
