// @inworld/nodejs-sdk/build/src/entities/scene.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scene = void 0;
const character_entity_1 = require("./character.entity");
class Scene {
    constructor(props) {
        this.characters = [];
        this.characters = props.characters;
        this.key = props.key;
    }
    static serialize(scene) {
        return JSON.stringify(scene);
    }
    static deserialize(json) {
        try {
            const { characters, key } = JSON.parse(json);
            return new Scene({ characters, key });
        }
        catch (e) { }
    }
    static fromProto(proto) {
        const characters = proto
            .getAgentsList()
            .map((agent) => {
            const assets = agent.getCharacterAssets();
            return new character_entity_1.Character({
                id: agent.getAgentId(),
                resourceName: agent.getBrainName(),
                displayName: agent.getGivenName(),
                assets: {
                    avatarImg: assets === null || assets === void 0 ? void 0 : assets.getAvatarImg(),
                    avatarImgOriginal: assets === null || assets === void 0 ? void 0 : assets.getAvatarImgOriginal(),
                    rpmModelUri: assets === null || assets === void 0 ? void 0 : assets.getRpmModelUri(),
                    rpmImageUriPortrait: assets === null || assets === void 0 ? void 0 : assets.getRpmImageUriPortrait(),
                    rpmImageUriPosture: assets === null || assets === void 0 ? void 0 : assets.getRpmImageUriPosture(),
                },
            });
        });
        return new Scene({
            key: proto.getKey(),
            characters,
        });
    }
}
exports.Scene = Scene;
