// @inworld/nodejs-sdk/build/src/entities/session_token.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionToken = void 0;
const TIME_DIFF_MS = 50 * 60 * 1000; // 5 minutes
class SessionToken {
    constructor(props) {
        this.token = props.token;
        this.type = props.type;
        this.expirationTime = props.expirationTime;
        this.sessionId = props.sessionId;
    }
    static isExpired(token) {
        const expirationTime = token.expirationTime;
        return (new Date(expirationTime).getTime() - new Date().getTime() <= TIME_DIFF_MS);
    }
    static serialize(token) {
        return JSON.stringify(Object.assign(Object.assign({}, token), { expirationTime: token.expirationTime.toISOString() }));
    }
    static deserialize(json) {
        try {
            const { token, type, expirationTime, sessionId } = JSON.parse(json);
            return new SessionToken({
                token,
                type,
                sessionId,
                expirationTime: new Date(expirationTime),
            });
        }
        catch (e) { }
    }
    static fromProto(proto) {
        return new SessionToken({
            token: proto.getToken(),
            type: proto.getType(),
            expirationTime: proto.getExpirationTime().toDate(),
            sessionId: proto.getSessionId(),
        });
    }
}
exports.SessionToken = SessionToken;
